client/login/magic-login/request-login-email-form.jsx

```diff
@@ -0,0 +1,139 @@
+//@TODO move this to wp-login compoenent
+import { getCurrentUser } from 'state/current-user/selectors';
+
+class RequestLoginEmailForm extends React.Component {
+	onEmailAddressFieldChange = event => {
+		this.props.setInputEmailAddress( event.target.value );
+	};
+
+	onSubmitBound = event => {
+		event.preventDefault();
+		this.props.onSubmit( this.props.emailAddress );
+	};
+
+	render() {
+		const {
+			currentUser,
+			requestError,
+			isFetching,
+			isValidEmailAddress,
+			onNoticeDismiss,
+			onClickEnterPasswordInstead,
+			emailRequested,
+			translate,
+		} = this.props;
+
+		const submitEnabled = (
+			isValidEmailAddress &&
+			! isFetching &&
+			! emailRequested &&
+			! requestError
+		);
+
+		const errorText = typeof requestError === 'string' && requestError.length
+			? requestError : translate( 'Unable to complete request' );
+
+		return (
+			<div>
+				{ requestError &&
+					<Notice
+						duration={ 10000 }
+						text={ errorText }
+						className="magic-login__request-login-email-form-notice"
+						showDismiss={ true }
+						onDismissClick={ onNoticeDismiss }
+						status="is-error" />
+				}
+				<LoggedOutForm onSubmit={ this.onSubmitBound }>
+					<p>{
+						translate( 'Get a link sent to the email address associated ' +
+							'with your account to log in instantly without your password.' )
+					}</p>
+					{ currentUser && currentUser.username &&
+						<p>{
+							translate( 'NOTE: You are already logged in as user: %(user)s', {
+								args: {
+									user: currentUser.username
+								}
+							} ) }</p>
+					}
+					<FormFieldset>
+						<FormTextInput
+							autoCapitalize="off"
+							autoFocus="true"
+							name="emailAddress"
+							placeholder="Email address"
+							value={ this.props.emailAddress }
+							onChange={ this.onEmailAddressFieldChange }
+						/>
+
+						<div className="magic-login__form-action">
+							<FormButton primary disabled={ ! submitEnabled }>
+								{ translate( 'Request Email' ) }
+							</FormButton>
+						</div>
+					</FormFieldset>
+				</LoggedOutForm>
+				<LoggedOutFormLinks>
+					<LoggedOutFormLinkItem onClick={ onClickEnterPasswordInstead }>
+						{ translate( 'Enter a password instead' ) }
+					</LoggedOutFormLinkItem>
+				</LoggedOutFormLinks>
+			</div>
+		);
+	}
+}
+
+const mapState = state => {
+	return {
+		currentUser: getCurrentUser( state ),
+		emailAddress: emailAddressFormInput( state ),
+		isFetching: isFetchingEmail( state ),
+		isValidEmailAddress: emailAddressFormInputIsValid( state ),
+		requestError: requestEmailError( state ),
+		emailRequested: requestedEmailSuccessfully( state ),
+	};
+};
+
+const mapDispatch = dispatch => {
+	return {
+		onClickEnterPasswordInstead: () => dispatch( hideRequestForm() ),
+		onNoticeDismiss: () => dispatch( hideRequestNotice() ),
+		onSubmit: emailAddress => dispatch( fetchRequestEmail( emailAddress ) ),
+		setInputEmailAddress: emailAddress => dispatch( setInputEmailAddress( emailAddress ) ),
```

Comment:
Instead of manually wrapping functions with dispatch, I recommend defining `mapDispatch` as an object:

```js
{
  onClickEnterPasswordInstead: hideRequestForm,
  …
}
``` 

It's shorter, less error-prone, and offloads mapper generation to Redux by being more declarative — allowing it to memoize if able to.